AI content updates (#17994)
ethereum/ethereum-org-website: public/content/ai-agents/wallets/index.md

---
title: AI agent wallets
description: Smart account and wallet infrastructure for Ethereum AI agents, including ERC-4337, EIP-7702, session keys, key management security patterns, and SDK guidance for ZeroDev and Safe
lang: en
template: ai-agents
emoji: ":purse:"
sidebarDepth: 2
image: /images/ai-agents/hero-image.png
alt: AI agent wallets on Ethereum

> summaryPoint1: Deploy smart account wallets with session keys, spending limits, and contract allowlists for safe agent autonomy

Reviewer: Dropped "and contract allowlists" — implied by the page; can keep if you prefer the explicit list.

```suggestion
summaryPoint1: Smart accounts with session keys and spending limits
```

> summaryPoint1: Deploy smart account wallets with session keys, spending limits, and contract allowlists for safe agent autonomy

Reviewer: Dropped "and contract allowlists" — implied by the page; can keep if you prefer the explicit list.

```suggestion
summaryPoint1: Smart accounts with session keys and spending limits
```

> summaryPoint2: Choose between ERC-4337 for new deployments and EIP-7702 for upgrading existing wallets

Reviewer: Trim to ≤8 words; keep keywords; parallel structure.

```suggestion
summaryPoint2: ERC-4337 for new wallets, EIP-7702 for upgrades
```

> summaryPoint2: Choose between ERC-4337 for new deployments and EIP-7702 for upgrading existing wallets

Reviewer: Trim to ≤8 words; keep keywords; parallel structure.

```suggestion
summaryPoint2: ERC-4337 for new wallets, EIP-7702 for upgrades
```

> summaryPoint3: Production key management patterns with human-in-the-loop escalation and spending policy enforcement

Reviewer: Dropped "spending policy enforcement" — covered by bullet 1.

```suggestion
summaryPoint3: Key management patterns with human-in-the-loop escalation
```

> summaryPoint3: Production key management patterns with human-in-the-loop escalation and spending policy enforcement

Reviewer: Dropped "spending policy enforcement" — covered by bullet 1.

```suggestion
summaryPoint3: Key management patterns with human-in-the-loop escalation
```

The proposed [Request Permissions from Wallets - ERC-7715](https://eips.ethereum.org/EIPS/eip-7715) addresses the lack of a unified interface by defining a standardized `wallet_grantPermissions` RPC method, through which an agent requests a scoped session key from a wallet provider. The policy parameters described above (spend limits, contract allowlist, validity window) are the values passed in that request. 

ERC-7715 is in draft and being adopted by major smart wallet providers. Until it is finalized, consult your chosen SDK's documentation for its current permission request API.

## SDK implementation examples {#sdk-examples}

The examples below cover two production-grade options that serve different deployment shapes:
- ZeroDev for new agent deployments requiring granular policy enforcement
- Safe for high-value or multi-party treasury deployments

Both have well-documented session key permission APIs. If evaluating other wallet SDKs, verify they provide stable, documented methods for enforcing session key boundaries (such as spend limits and contract allowlists) while ERC-7715 remains in draft. 

### ZeroDev (Kernel) {#zerodev-kernel}

> <Alert variant="info" className="my-8">

Reviewer: This `Best for:` block is a wallet-decision summary. Readers scan it to compare ZeroDev against the other options. It carries decision weight, not background context, so it should stand out from prose.

```suggestion
<Alert variant="warning" className="my-8">
```

  validUntil: Math.floor(Date.now() / 1000) + 60 * 60 * 24, // 24 hours
})

const permissionValidator = await toPermissionValidator(publicClient, {
  entryPoint: ENTRY_POINT_ADDRESS,
  signer: agentSigner, // agent's key — not the owner's
  policies: [callPolicy, timestampPolicy],
})

const agentAccount = await createKernelAccount(publicClient, {
  plugins: {
    sudo: ecdsaValidator,        // owner's public key (ECDSA address) — can revoke
    regular: permissionValidator, // agent's bounded authority
  },
  entryPoint: ENTRY_POINT_ADDRESS,
})

// Pass agentKernelClient (not the owner's client) to your framework wallet adapter
export const agentKernelClient = createKernelAccountClient({
  account: agentAccount,
  chain: sepolia,
  bundlerTransport: http(process.env.BUNDLER_RPC),
  middleware: {
    sponsorUserOperation: (await createZeroDevPaymasterClient({
      chain: sepolia,
      transport: http(process.env.PAYMASTER_RPC),
    })).sponsorUserOperation,
  },
})
```

Any UserOperation that violates `callPolicy` or `timestampPolicy` is rejected by the bundler before it reaches the mempool.

### Safe (Allowance Module) {#safe-allowance-module}

> <Alert variant="info" className="my-8">

Reviewer: Same as the ZeroDev `Best for:` above. Safe's decision-summary should match the visual weight of the other wallet options on the page.

```suggestion
<Alert variant="warning" className="my-8">
```
